### PR TITLE
Max score

### PR DESCRIPTION
**Pull request by tsivert**
This is a portion of the logic described here: https://h5ptechnology.atlassian.net/browse/HFP-1806
The rest is in the pull request for h5p/h5p-editor-php-library.

Added support for loading the presave file if exist in a library.

### CODE REVIEW

File: h5p-default-storage.class.php [hunk 1]
```diff
@@ -451,6 +451,18 @@ public function hasWriteAccess() {
     return self::dirReady($this->path);
   }
 
+    /**
+     * Check if the file presave.js exists in the root of the library
+     *
+     * @param string $name
+     * @return bool
+     */
+  public function hasPresave($libraryFolder, $developmentPath = null) {
```
Comment (thomasmars) [inline h5p-default-storage.class.php]: specify params in phpdoc

File: h5p-file-storage.interface.php [hunk 2]
```diff
@@ -190,4 +190,11 @@ public function removeContentFile($file, $contentId);
    * @return bool True if server has the proper write access
    */
   public function hasWriteAccess();
+
+  /**
+   * Check if the library has a presave.js in the root folder
+   *
+   * @return string|null Path if presave.js found
+   */
+  public function hasPresave($libraryName, $developmentPath = null);
```
Comment (thomasmars) [inline h5p-file-storage.interface.php]: specify params in phpdoc

File: h5p.classes.php [hunk 3]
```diff
@@ -2130,6 +2134,26 @@ public function getDependenciesFiles($dependencies, $prefix = '') {
     return $files;
   }
 
+    public function hasPresave($libraryName)
+    {
+        if( isset($this->h5pD) ){
+            extract(H5PCore::libraryFromString($libraryName));
```
Comment (thomasmars) [inline h5p.classes.php]: Please don't use extract, it makes it harder to follow the code and may expose unintended variables as well as making static code analysis much harder. Especially when the function the values are retrieved from is not properly documented.

File: h5p.classes.php [hunk 4]
```diff
@@ -2130,6 +2134,26 @@ public function getDependenciesFiles($dependencies, $prefix = '') {
     return $files;
   }
 
+    public function hasPresave($libraryName)
+    {
```
Comment (thomasmars) [inline h5p.classes.php]: brackets should be on same line as function and indentations should be 2 spaces

File: h5p.classes.php [hunk 5]
```diff
@@ -2130,6 +2134,26 @@ public function getDependenciesFiles($dependencies, $prefix = '') {
...
+            $library = $this->h5pD->getLibrary($machineName, $majorVersion, $minorVersion);
+            if( !is_null($library)){
+                return $this->fs->hasPresave($libraryName, $library['path']);
+            }
+        }
+        return $this->fs->hasPresave($libraryName);
+    }
+
+  public function addPresaveFile(&$assets, $library, $prefix = ''){
+    $this->getDependencyAssets([
```
Comment (thomasmars) [inline h5p.classes.php]: try to keep line length to a max of 80 chars, and split the values up if they're longer.

**Comment by tsivert** [comment]
Thanks @thomasmars  for your feedback. I'll fix the issues and get back to you.

An example implementation for a framework can be view here: https://github.com/h5p/h5p-mark-the-words/pull/14. For more examples take a look at all my forked repositories of your frameworks. :)

**Comment by tsivert** [comment]
Hi again.
I can't reproduce the part of your feedback where presave.js is loaded multiple times. Can you point me in a direction on how to do that?

**Comment by tsivert** [comment]
I've now changed the code so that presave.js only loads when editing a H5P. But then I remembered that @falcon-git mentioned in an email that it was supposed to load from the h5p-php-library(when viewing a H5P) in order to be extended this later for other purposes. Should I then revert the changes I just did so that it loads in both places or keep it as is(only load when editing)?

**Comment by thomasmars** [comment]
I don't see any reason why it would load from h5p-php-library. @falcon-git has to chime in if he thinks differently. The main purpose of this feature and its naming scheme suggests that it only should run before the content is saved, this only happens in the editor. If the functionality is needed in the view at some later point, then it should be handled at that time in my opinion. There's no need for you to revert your changes.
Were you able to see when Course Presentation presave.js was loaded multiple times ? It happens for the example content of Course Presentation, perhaps it was fixed by your latest changes that prevents presave.js from loading on view.

**Comment by tsivert** [comment]
No I was not able to reproduce. I added a console.log at the top of the presave.js file for the CP to see how many times it was loaded. And the console displayed only one instance of the debug text. 
But the presave file for each library used in CP that supports calculation of max score is also loaded when editing a CP. Could it be those presave files you are seeing?

**Comment by tsivert** [comment]
The issue with loading multiple times was fixed when I removed the loading of presave.js from the view.

File: h5p-default-storage.class.php [hunk 6]
```diff
@@ -451,6 +451,19 @@ public function hasWriteAccess() {
...
 
+    /**
+     * Check if the file presave.js exists in the root of the library
+     *
+     * @param string $libraryFolder
+     * @param string $developmentPath
+     * @return bool
+     */
+  public function hasPresave($libraryFolder, $developmentPath = null) {
+      $path = is_null($developmentPath) ? 'libraries' . DIRECTORY_SEPARATOR . $libraryFolder : $developmentPath;
```
Comment (thomasmars) [inline h5p-default-storage.class.php]: Please fix indentation issues here

File: h5p.classes.php [hunk 7]
```diff
@@ -2109,8 +2109,9 @@ public function getDependenciesFiles($dependencies, $prefix = '') {
 
     // Using content dependencies
     foreach ($dependencies as $dependency) {
+      $libraryName = H5PCore::libraryToString($dependency, TRUE);
```
Comment (thomasmars) [inline h5p.classes.php]: There shouldn't be any necessary changes in h5p.classes.php anymore, right ?

**Pull request merged**